stdlib/src/collections/list.mojo

```diff
@@ -858,6 +858,64 @@ struct List[T: CollectionElement](CollectionElement, Sized, Boolable):
         """
         return self.data
 
+    @always_inline
+    fn __eq__[
+        Rhs_t: EqualityCollectionElement, //
+    ](self: Self, rhs: List[Rhs_t]) -> Bool:
+        """Compare this List to the RHS using EQ comparison.
+
+        Note that it is not yet possible to compare lists of lists.
+
+        Args:
+            rhs: The other List to compare against.
+
+        Parameters:
+            Rhs_t: The inferred type of rhs.T.
+
+        Returns:
+            True if this List is equal to the RHS List and False otherwise.
+        """
+
+        @parameter
+        if not _type_is_eq[T, Rhs_t]():
+            return False
```

Comment:
Isn't this situation always a user error? 

As a user, if I'm comparing two list of struct with different types, I would most of the time want the compiler to tell me at compile-time since the result is always the same. 

Maybe let's disallow this, it will remove the need for a `rebind`, that will make this PR easier to merge, and if later on, we see the need for this function to check for equality of the types, we can add it. I would advocate for the "wait-and-see" approach.

Looking at mypy, they provide a command line argument to disallow this:
https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-strict-equality

if enabled, you get the following error:

```
trying_stuff.py:5: error: Non-overlapping equality check (left operand type: "list[int]", right operand type: "list[str]")  [comparison-overlap]
Found 1 error in 1 file (checked 1 source file)
```
@rd4com what do you think about this solution?

Reply:
I think we should use the new parametric self to implement this.

Reply:
@soraros , @laszlokindrat , @JoeLoser , @gabrieldemarmiesse 
I messed up the git tree and tried everything:
feel free to make a PR by using this one, i give up this one.